From 9584e39b3a024269adda4d089ecd115657368a8e Mon Sep 17 00:00:00 2001
From: huzhen <george.hu@spacemit.com>
Date: Fri, 5 Jul 2024 11:27:01 +0800
Subject: [PATCH 1022/1448] rtl8852be: Fix a problem might cause kernel panic

  GET_PHL_INFO(_dvobj)	is defined (_dvobj->phl),
  dvobj and dvobj->phl would be NULL.

Change-Id: Ib3697d4f918d9f0015a3f867815f43052e33f739
---
 drivers/net/wireless/realtek/rtl8852be/core/rtw_phl.c | 6 ++++--
 drivers/net/wireless/realtek/rtl8852be/phl/phl_role.c | 8 +++++++-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/realtek/rtl8852be/core/rtw_phl.c b/drivers/net/wireless/realtek/rtl8852be/core/rtw_phl.c
index 756d33873a3c..c619187c78ee 100644
--- a/drivers/net/wireless/realtek/rtl8852be/core/rtw_phl.c
+++ b/drivers/net/wireless/realtek/rtl8852be/core/rtw_phl.c
@@ -1481,8 +1481,10 @@ void rtw_hw_iface_deinit(_adapter *adapter)
 	rtw_phl_ps_set_rt_cap(GET_PHL_INFO(dvobj), HW_BAND_0, ps_allow, PS_RT_CORE_INIT);
 #endif
 	if (adapter->phl_role) {
-		rtw_free_self_stainfo(adapter);
-		rtw_phl_wifi_role_free(GET_PHL_INFO(dvobj), adapter->phl_role->id);
++		if (dvobj && dvobj->phl) {
++			rtw_free_self_stainfo(adapter);
++			rtw_phl_wifi_role_free(GET_PHL_INFO(dvobj), adapter->phl_role->id);
++		}
 		adapter->phl_role = NULL;
 	}
 #if defined(CONFIG_RTW_IPS) || defined(CONFIG_RTW_LPS)
diff --git a/drivers/net/wireless/realtek/rtl8852be/phl/phl_role.c b/drivers/net/wireless/realtek/rtl8852be/phl/phl_role.c
index 1a26f2f20fb5..3c90a08a5c6e 100644
--- a/drivers/net/wireless/realtek/rtl8852be/phl/phl_role.c
+++ b/drivers/net/wireless/realtek/rtl8852be/phl/phl_role.c
@@ -1789,7 +1789,7 @@ phl_wifi_role_stop(struct phl_info_t *phl_info, struct rtw_wifi_role_t *wrole)
 void rtw_phl_wifi_role_free(void *phl, u8 role_idx)
 {
 	struct phl_info_t *phl_info = (struct phl_info_t *)phl;
-	struct rtw_phl_com_t *phl_com = phl_info->phl_com;
+	struct rtw_phl_com_t *phl_com = NULL;
 	struct rtw_wifi_role_t *wrole = NULL;
 
 	if (role_idx >= MAX_WIFI_ROLE_NUMBER) {
@@ -1797,6 +1797,12 @@ void rtw_phl_wifi_role_free(void *phl, u8 role_idx)
 		return;
 	}
 
+	if (!phl_info || !phl_info->phl_com) {
+		PHL_ERR("%s phl_info or phl_com is NULL\n", __func__);
+		return ;
+	}
+
+	phl_com = phl_info->phl_com;
 	wrole = &phl_com->wifi_roles[role_idx];
 
 	if (phl_wifi_role_stop(phl_info, wrole) != RTW_PHL_STATUS_SUCCESS) {
-- 
2.47.0

